### PR TITLE
IO non-buffered read optimization

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
This PR implements an optimization that avoids buffering in the QIO channel for large read operations. Buffering is typically useful for a series of small reads, as it avoids the latency/overhead associated with a large number of system calls; however, for sufficiently large read operations, buffering imposes unneeded overhead so it can be avoided.

This optimization only takes effect when:

* the number of bytes to be written is larger than the threshold specified by qio_write_unbuffered_threshold
* the channel's method is read/write, fread/fwrite, or pread/pwrite (no mmap)
* there are no pending commits/reverts (the mark stack only contains "right_mark_start")
* there is no IO plugin associated with the channel

Tests are included to validate the correctness of this change.

- [x] paratest

### CODE REVIEW

File: runtime/include/qio/qio.h [hunk 1]
```diff
@@ -1040,6 +1041,7 @@ qioerr qio_channel_read_amt(const int threadsafe, qio_channel_t* restrict ch, vo
   if( qio_space_in_ptr_diff(len, ch->cached_end, ch->cached_cur) ) {
     qio_memcpy( ptr, ch->cached_cur, len );
     ch->cached_cur = qio_ptr_add(ch->cached_cur, len);
+    ch->mark_stack[ch->mark_cur] += len;
```
Comment (mppf) [inline runtime/include/qio/qio.h:1044]: I don't think this change is needed. The concept is that, when `ch->cached_cur` and `ch->cached_end` are in use, the current position is represented by `ch->cached_cur` itself. See e.g. `qio_channel_offset_unlocked` which shows how the offest is computed in this manner.

But, presumably there was some reason you added it. Could you address that concern in another way?

File: runtime/src/qio/qio.c [hunk 2]
```diff
@@ -2798,56 +2802,140 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
...
+    // this way, the buffer is ready for any future buffered reads
+    // (the buffer is empty, so we can just move the end pointers)
+    qbuffer_reposition(&ch->buf, qbuffer_start_offset(&ch->buf) +
+                       *amt_read - gotlen);
+    ch->av_end += *amt_read - gotlen;
 
-    // move the iterator for the next read
-    qbuffer_iter_advance(&ch->buf, &start, gotlen);
+    // re-setup the buffer for any future buffered writes
+    _qio_buffered_setup_cached(ch);
```
Comment (mppf) [inline runtime/src/qio/qio.c:2896]: Now that you have some performance testing, you can answer the question, is it faster if you leave this line out?

**Review by mppf — APPROVED** [review]

File: runtime/include/qio/qio.h (hunk 1 above)
Comment (jeremiah-corrado) [reply runtime/include/qio/qio.h:1044]: Without this line, the next read was not picking up where the previous had left off (it seemed like `cached_cur` was not being used to represent the channel's position in the file). 

I'll try to figure out why and hopefully solve this problem in a more elegant way.

Comment (jeremiah-corrado) [reply runtime/include/qio/qio.h]: Good point, that has been fixed now.

**Pull request merged**